### DISCUSSION TITLE
Oh-my-posh hiding `[DBG]` prefix in powershell.

### DISCUSSION THREAD

**Discussion by awa5114** [JanDeDobbeleer/oh-my-posh#1928] [Troubleshoot]
I am trying to debug a powershell script with Windows Terminal. I have oh-my-posh installed with a customized command prompt. I went through the following steps:

 - `set-psbreakpoint -script test.ps1 -line 8`
 - .\test.ps1

I got the following output:

```
Hit Line breakpoint on 'test.ps1:8'

At test.ps1:8 char:3
+   cat $py_file.fullname
+   ~~~~~~~~~~~~~~~~~~~~~
```

Which implies that my debugger is working fine. Furthermore I'm able to print out variables from within `test.ps1`.

Unfortunately, it looks like oh-my-posh is hiding the `[DBG]` prefix in my command line prompt, which I would like to have.

How can I prevent oh-my-posh from hiding the `[DBG]` prefix in my command line prompt?

**Comment by JanDeDobbeleer** [comment]
That's not something we support just yet, but let me see how this works and what needs to be done to enable that.

**Reply by JanDeDobbeleer** [reply]
This is handled by the following variable:

```powershell
function prompt {
    $(if (Test-Path variable:/PSDebugContext) { '[DBG]: ' }
      else { '' }) + 'PS ' + $(Get-Location) +
        $(if ($NestedPromptLevel -ge 1) { '>>' }) + '> '
}
```

**Reply by JanDeDobbeleer** [reply]
You can already make use of the following functionality to add a text segment that indicates the right information. I will however also create additional functionality to display a smaller prompt.

```powershell
function Set-EnvVar {
    $env:POSH_DEBUG=Test-Path variable:/PSDebugContext
}
New-Alias -Name 'Set-PoshContext' -Value 'Set-EnvVar' -Scope Global -Force
```

Use this context like this:

```json
{
  "type": "text",
  "style": "plain",
  "foreground": "#E06C75",
  "properties": {
    "template": "{{ if eq .Env.POSH_DEBUG \"True\" }} [DBG] {{ end }}"
  }
}
```

To validate if the if statement is correct, I didn't check that.

**Reply by mattcargile** [reply]
Thanks for the sample code, Jan! Thanks for the great idea @awa5114. 

Unfortunately, I couldn't get this to work in `pwsh.exe` or `powershell.exe` . It appears the process overrides the `function:\prompt`. When running `ls Function:\prompt | select -ExpandProperty Definition` in the `$PSDebugContext`, the `function` doesn't change. So something internal must be changing it? Here is a example below.
```powershell
ni t.ps1 -value '''hello world'''
Set-PSBreakpoint -Script .\t.ps1 -Line 1
./t.ps1
```
![image](https://user-images.githubusercontent.com/74641867/160260665-aa940243-c99d-4b69-a509-8e9395288baf.png)

I am using `wt.exe` and Powershell `7.2.2` and Powershell `5.1`.

**Reply by JanDeDobbeleer** [reply]
@mattcargile yup, there's now the debug prompt (just like secondary). But I didn't document it yet 🤦🏻‍♂️ Follow the secondary prompt [guide](https://ohmyposh.dev/docs/config-secondary-prompt) but use `debug_prompt` instead of `secondary_prompt` as name.

**Reply by mattcargile** [reply]
@JanDeDobbeleer , I still can't get this to work for some reason? I see the `Test-Path variable:/PSDebugContext` in my `function:\prompt` but I just can't get it to work in VS Code or `pwsh.exe` in `wt.exe`.

**Reply by mattcargile** [reply]
After playing with it more, I'm not sure `$PSDebugContext` is available within `function:\prompt` for some reason or maybe it is the `New-Module` wrapper. `Test-Path Variable:/PSDebugContext` was returning false when I embedded `Write-Host` into my `function:\prompt`.

**Reply by mattcargile** [reply]
Looks like VSCode has an upcoming fix [here](https://github.com/PowerShell/vscode-powershell/issues/3980) for it. I don't know why it won't work in `pwsh.exe` though.

**Reply by mattcargile** [reply]
I'm assuming the `function:\prompt` would need to be brought outside the `oh-my-posh-core` module including the associated functions. Should I make a issue for it?

**Reply by mattcargile** [reply]
Yeah I just tested it and removing the module wrapper fixes the debug deal.

**Reply by mattcargile** [reply]
Now I can get it to change the prompt but it won't output the `debug_prompt` config. It only outputs `PS> ` as the prompt.

**Reply by JanDeDobbeleer** [reply]
@mattcargile isn't there another way to get that context than `Test-Path Variable:/PSDebugContext`. I don't see why it being inside the module would behave differently unless PSReadLine has an issue?

**Reply by mattcargile** [reply]
Hmmm. There is some kind of special debug context that is preventing the module from seeing it, I guess. I couldn't get it to work. Maybe there is another variable to use. The official docs has an example that uses that variable.

**Reply by lewis-yeung** [reply]
After removing the module wrapper, I tested the `prompt` function in a debug context, and it threw a `MethodInvocationException`:

```
Exception calling "Start" with "0" argument(s): "Cannot start process because a file name has not been provided."
```

I found that `$script:OMPExecutable` showed nothing, while `$OMPExecutable` worked. The `prompt` function failed and that's why only a `PS>` is output as the default prompt.

Furthermore, all variables defined in the `script` scope become unavailable in a debug context. There seems to be a special "isolated scope" when using a module wrapper even though I have exported the functions and imported the module with `-Global`. After removing all `script:` prefixes, everything works, of course, including the `debug_prompt`. 😅

PowerShell has got these things tricky. Not sure if I could find a better solution. 😭 Any ideas, friends?

**Reply by lewis-yeung** [reply]
Figured out a simple fix just now. Add a `Test-PSDebugContext` function outside the module wrapper and call it in `prompt`:

```powershell
function Test-PSDebugContext { return Test-Path Variable:/PSDebugContext }
```

It works in PowerShell 7.2.4 with either Windows Terminal or VSCode on Win11. See #2294. @JanDeDobbeleer @mattcargile

**EDITED:** It works in Windows PowerShell 5.1 as well.

**Reply by mattcargile** [reply]
Awesome! Can't wait to try it. I assumed it was some debug context deal. Hopefully we can get this in the repo soon if it is reasonable.

**Reply by lewis-yeung** [reply]
@mattcargile It's live now. You can try it and if anything is still wrong, just leave a comment.

**Reply by JustinGrote** [reply]
@lewis-yeung  @mattcargile does this still work? I can't seem to get debug_prompt to work correctly, it shows my same prompt when going into debugger.

**Reply by mattcargile** [reply]
@JustinGrote - It has worked in _most_ contexts, though admittedly I haven't tried it super lately on v12 or later. 

I have noticed it breaking depending on how the module is loaded or how one sets the breakpoint like on the command itself versus on a line. I've been meaning to put in an Issue, but I haven't been able to get to it. One needs additional logic on `Get-PSCallStack` output during various debug contexts, I _think_.

**Reply by JustinGrote** [reply]
@mattcargile I recently got a PR merged to handle the other contexts such as ephemeral that weren't working for me. Should be pretty solid now for 99% of use cases.
